qos/evm/check_chainid.go

```diff
@@ -0,0 +1,55 @@
+package evm
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/buildwithgrove/path/qos/jsonrpc"
+)
+
+// methodChainID is the JSON-RPC method for getting the chain ID.
+// Reference: https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_chainid
+const methodChainID = jsonrpc.Method("eth_chainId")
+
+// TODO_IN_THIS_PR(@commoddity): determine an appropriate interval for checking the chain ID.
+const checkChainIDInterval = 20 * time.Minute
+
+var (
+	errNoChainIDObs      = fmt.Errorf("endpoint has not had an observation of its response to a %q request", methodChainID)
+	errInvalidChainIDObs = fmt.Errorf("endpoint returned an invalid response to a %q request", methodChainID)
+)
+
+// endpointCheckChainID is a check that ensures the endpoint's chain ID is the same as the expected chain ID.
+// It is used to ensure that the endpoint is on the correct chain.
+type endpointCheckChainID struct {
+	// chainIDResponse stores the result of processing the endpoint's response to an `eth_chainId` request.
+	// It is nil if there has NOT been an observation of the endpoint's response to an `eth_chainId` request.
+	chainID   *string
+	expiresAt time.Time
+}
+
+// isValid returns an error if the endpoint's chain ID does not match the expected chain ID in the service state.
+func (e *endpointCheckChainID) isValid(evmChainID string) error {
```

Comment:
Ditto re logic consolidation: great idea, needs better placement.

Reply:
I've moved this logic to the service state.